README.md
[![NPM](https://nodei.co/npm/yocto-config.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/yocto-config/)

![alt text](https://david-dm.org/yoctore/yocto-config.svg "Dependencies Status")
[![Code Climate](https://codeclimate.com/github/yoctore/yocto-config/badges/gpa.svg)](https://codeclimate.com/github/yoctore/yocto-config)
[![Test Coverage](https://codeclimate.com/github/yoctore/yocto-config/badges/coverage.svg)](https://codeclimate.com/github/yoctore/yocto-config/coverage)
[![Issue Count](https://codeclimate.com/github/yoctore/yocto-config/badges/issue_count.svg)](https://codeclimate.com/github/yoctore/yocto-config)
[![Build Status](https://travis-ci.org/yoctore/yocto-config.svg?branch=master)](https://travis-ci.org/yoctore/yocto-config)

## Overview

This module is a part of yocto node modules for NodeJS.

Please see [our NPM repository](https://www.npmjs.com/~yocto) for complete list of available tools (completed day after day).

This module provide a simple config validator tools for a node app.

## Motivation

Our main motivation for this module is, create and provide a single, simple and custom config manager (with validation) for each part of our program in the same place.

## How validation schema works ?

Default validation schema was build with [joi](https://www.npmjs.com/package/joi) package manager, and all custom all schema must be associate with a **joi rules too**.

During load process given files was auto validated with associated schema.


## Installation

```bash
npm install --save yocto-config
```

## Config files priority

**Priority** : `(Other file).json` < `all.json` < `common.json` < `development.json` < `stagging.json` < `production.json`

all.json : place in this file all general property (shared key, node port, etc) 

common.json : place in this file all common data between each env (app name, express config, etc) 

development.json : place in this file all development property for development environnement

stagging.json : place in this file all stagging property for stagging environnement

production.json : place in this file all production property for production environnement

**IMPORTANT** : file was merged with previous defined priority, so it should be understood that we dont need to define multiple times the same property if we doesn't need to replace it. **JUST PLACE IN CORRECT ENV WHAT YOU NEED**

*For example :* 

```javascript
// a development.json file with these property
{
"test" : {
  "db" : {
    "uri" : "http://test.com/123456"
  }
}
```

```javascript
// a production.json file with these property
{
"test" : {
  "db" : {
    "options" : {
        "op1" : "my-value1",
        "op2" : "my-value2"
    }
  }
}
```

Will produce on production this config data :
```javascript
// generated config in production env
{
"test" : {
  "db" : {
    "uri" : "http://test.com/123456",
    "options" : {
        "op1" : "my-value1",
        "op2" : "my-value2"
    }
  }
}
```

## Pre-defined configuration

Predefined configuration schema was availabe. To use them see methods below : 

- [enableExpress]() : add to default schema an express configuration
- [enableMongoose]() : add to default schema a mongoose configuration
- [enablePassportJs]() : add to default schema a passportjs configuration for (facebook, twitter, google, active-directory, standard login)
- [enableRender]() : add to default schema a [yocto-render](https://www.npmjs.com/package/yocto-render) configuration
- [enableRouter]() : add to default schema a [yocto-router](https://www.npmjs.com/package/yocto-router) configuration

All of these function was replaced by default an already defined configuration.

If you want to enable a new configuration and keep safe the previous configuration, just pass `true` on enable function, and the new config will be append on the previous content, for example :

```javascript
var config  = require('yocto-config')();

// defined base config
config.enableExpress();
// add new config for mongoose and keep safe previous defined config
config.enableMongoose(true); 
```

## Adding custom config schema

You can add your custom config schema for custom validation, two methods was available for these action : 

- addCustomSchema(name, schema, autoenable, keepSafePreviousSchema) : add new schema with given name
- enableSchema(name, keepSafePreviousSchema) : enable a schema with given name

Example :

```javascript
var config  = require('yocto-config')();
var joi     = require('joi');

// define your schema
var schema = joi.object().required().keys({
  mailer  : joi.object().required(),
  daemon  : joi.object().required(),
  core    : joi.object().keys({
    daemon : joi.object().default({ delay : 10, retry : 10 }).keys({
      delay : joi.number().min(0).default(10),
      retry : joi.number().min(0).default(10)
    }).allow([ 'delay', 'retry' ]),
    mailer : joi.object().default({ limit : 1, sort : 'ascending' }).keys({
      limit : joi.number().min(1).default(1),
      sort  : joi.string().empty().default('ascending').allow([ 'ascending', 'descending' ]),
    }).allow([ 'limit', 'sort' ])
  }).allow([ 'daemon', 'mailer' ])
}).unknown();

// add custom schema named 'test'
// third parameter enable given schema (optional, by default to false)
// fourth parameter keep safe all predefined schema (optional, by default to false)
config.addCustomSchema('test', schema, true, true);

// OR classic process
config.addCustomSchema('test', schema);
config.enableSchema('test');
```

## How to use

```javascript

var config  = require('yocto-config')();

// enable defined config
config.enableExpress(true);
config.enableMongoose(true);
config.enablePassportJs(true);

// Set config path
config.setConfigPath('./example/config');

// Load and retreive data from callback
config.load().then(function(data) {
  // your code here
  // you can also get data with a method
  var c = config.getConfig();
  console.log(c);
}).catch(function(error) {
  // your code here
});

```

## How to change config on the fly ?

In some case we need to use same core app for different apps. 
To change the config path for each app during run it's possible to use `process.env`.
To use it run during start your app use these params `CONFIG_SUFFIX_PATH`.

```bash
CONFIG_SUFFIX_PATH='suffix/base/path' node app
```

## Tricks

You can also use a utility method `autoEnableValidators` to enable your validator. See below key associated with schema : 

- `express` for express schema
- `passportJs` for passportJs schema
- `mongoose` for mongoose schema
- `render` for yocto-render module schema
- `router` for yocto-router module schema

Example : 

```javascript
var config  = require('yocto-config')();
config.autoEnableValidators([ 'express', 'passportJs', 'mongoose' ]);
// Extra process
```

## Default configuration rules

- Mongoose schema can be find [here](https://yoctore.github.io/yocto-config/schema.js.html#line741)
- Express schema can be find [here](https://yoctore.github.io/yocto-config/schema.js.html#line383)
- Yocto Render schema can be find [here](https://yoctore.github.io/yocto-config/schema.js.html#line869)
- Yocto Router schema can be find [here](https://yoctore.github.io/yocto-config/schema.js.html#line966)
- Yocto PassportJS schema can be find [here](https://yoctore.github.io/yocto-config/schema.js.html#line805)

## Logging in tool

By Default this module include [yocto-logger](https://www.npmjs.com/package/yocto-logger) for logging.
It's possible to inject in your config instance your current logger instance if is another `yocto-logger` instance.

## Changelog

All history is [here](https://github.com/yoctore/yocto-config/blob/master/CHANGELOG.md)

## Full Documenation

You can find full online documentation [here](https://yoctore.github.io/yocto-config)

CHANGELOG.md
## 3.0.0 (2018-02-14)

- Major update of all dependencies
- Update node version engine to >= 8.9.4

## 2.0.2 (2017-04-20)
- Add into express schema the object staticServe

## 2.0.0 (2016-10-07)

- Major package update dependencies
- Add new env value `CONFIG_SUFFIX_PATH` to append like a suffix on the end of config path
- Refactoring and merging all schema Class in a single class
- Adding documentation form yocto-doc
- Add more unit tests

## 1.8.2 (2016-10-07)

- Add config for fingerprint for render part

## 1.8.2 (2016-09-27)

- Update config and add new rules for seo redirection and www redirection process

## 1.8.1 (2016-09-14)

- Update version for github and npm

## 1.8.0 (2016-09-14)

- Add sandbox rules optional for chrome compatibility
- Add cors config from a `corsCfg` property

## 1.7.6 (2016-08-08)
- Can enable or disable security

## 1.7.5 (2016-06-28)

- Add property <autoEncryptRequest> into JWT object
- Add property <autoDecryptRequest> into JWT object

## 1.7.4 (2016-05-27)

- Add property <allowedRoute> into JWT object

## 1.7.3 (2016-03-23)

- Add options rules on session for SSL/TLS connexion.

## 1.6.2 (2015-12-28)

- Fix a bug on NODE_ENV usage

## 1.6.0 && 1.6.1 (2015-12-10)

- add support for HTTPS
- add new rules on express schema to manage protocol and cert files
- move port rule on protocol object rules

## 1.5.0 (2015-12-09)

- Change passportjs schema for activeDirectory and standard connection. move from object to an array of object. See doc for more details.
- Add documentation for passportJs usage

## 1.4.0 (2015-12-02)

- Add jwt ips filter on config file

## 1.3.9 (2015-11-27)

- Fix a bug with allow-from xframe rules

## 1.3.6 (2015-11-04)

- Adding angular rules on lusca for csrf rules

## 1.3.5 (2015-11-04)

- Add an allow filter on new render rules

## 1.3.4 (2015-11-04)

- Add rules for social item on render config

## 1.3.2 <-> 1.3.3 (2015-11-04)

- Update yocto dependencies package

## 1.3.1 (2015-11-03)

- Update package and jump to v7.0.0 for joi.

## 1.3.0 (2015-10-23)

- Add cors rules on default express schema.

## 1.2.0 (2015-10-23)

- Add [yocto-render](https://www.npmjs.com/package/yocto-render) config rules and `enableRender` method to enable this config
- Add [yocto-router](https://www.npmjs.com/package/yocto-router) config rules and `enableRouter` method to enable this config

## 1.1.1 (2015-10-23)

- Change level of log for find method
- Change use of logger on auto enable method

## 1.1.0 (2015-10-23)

- Change `passportjs` config key to `passportJs` (camelCase).
- Add a new method to auto enable method from given key
- Add a `type` property on mongoose schema.

## 1.0.6 (2015-10-22)

- Change message when no files was found.

## 1.0.4 (2015-10-16)

- Add jwt rules on default express config format

## 1.0.1 && 1.0.2 && 1.0.3 (2015-10-08)

- Change readme.md content & add new setConfigPath method
- Minor fixes
- Change directory config rules (all dir is not required now)

## 0.3.0 -> 1.0.0 (2015-10-06 <-> 2015-10-07)

- Add readme description
- Add rules to process enable and activate in the same method (1.0.0)
- Add a method to enable a specific schema (0.9.0)
- Add a method to add custom schema (0.8.0)
- Add access method (0.7.0) - to enable passportJs / express / mongoose config
- Add module passportJs (0.5.0) - default config for passportjs
- Add module express  (0.4.0) -  default config for express
- Add module mongoose  (0.3.0) -  default config for mongoose

## 0.2.1 (2015-10-06)

- Hint program with yocto-hitn and fix bugs
- Add Promise return (Q) on load function.

## 0.1.0 (2015-09-25)

Add all base function.
- load : load and test given configuration from given path
- set : default set function to set value
- get : default get value to get property value
- realod : reload configuration from given path if new path was given


src/index.js
'use strict';

var path    = require('path');
var _       = require('lodash');
var logger  = require('yocto-logger');
var joi     = require('joi');
var fs      = require('fs');
var glob    = require('glob');
var utils   = require('yocto-utils');
var Q       = require('q');
var schema  = require('./schema');

/**
 * Yocto config manager.
 * Manage your configuration file (all / common / env  & specific file)
 *
 * Config file has priority. And priority is defined like a php ini system.
 * `(Other file).json` < `all.json` < `common.json` < `development.json` < `stagging.json` < `production.json`
 *
 * All specific data must be configured on a each correct file.
 *
 * all.json : contains general data
 * common.json : must contains all common data between each env
 * development.json : must contains development data for development environnement
 * staging.json : must contains stagging data for staging environnement
 * production.json : must contains production data for production environnement
 *
 * This Module use some security format rules based on Lusca NPM module : {@link https://www.npmjs.com/package/lusca}
 * - Cross Site Request Forgery (CSRF) headers : {@link https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)}
 * - Content Security Policy (CSP) headers : {@link https://www.owasp.org/index.php/Content_Security_Policy}
 * - MDN CSP usage : {@link https://developer.mozilla.org/en-US/docs/Web/Security/CSP/Using_Content_Security_Policy}
 * - X-FRAME-OPTIONS : {@link https://www.owasp.org/index.php/Clickjacking}
 * - Platform for Privacy Preferences Project (P3P) headers : {@link http://support.microsoft.com/kb/290333}
 * - HTTP Strict Transport Security (HSTS & Chrome HSTS preload) : {@link https://www.owasp.org/index.php/HTTP_Strict_Transport_Security} & {@link https://hstspreload.appspot.com/}
 * - XssProtection : {@link http://blogs.msdn.com/b/ie/archive/2008/07/02/ie8-security-part-iv-the-xss-filter.aspx}
 *
 * @date : 12/05/2015
 * @author : ROBERT Mathieu <mathieu@yocto.re>
 * @copyright : Yocto SAS, All right reserved
 * @class Config
 * @param {Object} logger Defaut yocto-logger instance
 */
function Config (logger) {
  /**
   * Default config value
   *
   * @public
   * @memberof Config
   * @member {Object} config
   * @default {}
   */
  this.config = {};

  /**
   * Default state value, true if config state is or false otherwise
   *
   * @public
   * @memberof Config
   * @member {Boolean} state
   * @default false
   */
  this.state = false;

  /**
   * Default env value
   *
   * @public
   * @memberof Config
   * @member {String} env
   * @default development
   */
  this.env = process.env.NODE_ENV || 'development';

  /**
   * Default base path
   *
   * @public
   * @memberof Config
   * @member {String} base
   */
  this.base = process.cwd();

  /**
   * Default logger instance. can be override by set function
   *
   * @public
   * @memberof Config
   * @member {Instance} logger
   */
  this.logger = logger;

  /**
   * Prefix to use in case of multiple configuration
   *
   * @public
   * @memberof Config
   * @member {String} prefix
   */
  this.suffix = process.env.CONFIG_SUFFIX_PATH || '';

  /**
   * Default schema validation for config validator
   *
   * @public
   * @memberof Config
   * @member {Object}schema
   * @default {
   *  express       : schema.getExpress(),
   *  mongoose      : schema.getMongoose(),
   *  passportJs    : schema.getPassportJs(),
   *  render        : schema.getRender(),
   *  router        : schema.getRouter()
   * }
   */

  this.schemaList = {
    express    : schema.getExpress(),
    mongoose   : schema.getMongoose(),
    passportJs : schema.getPassportJs(),
    render     : schema.getRender(),
    router     : schema.getRouter()
  };

  /**
   * Current schema to use for validation
   *
   * @public
   * @memberof Config
   * @property {Object} schema
   * @default {}
   */
  this.schema = {};
}

/**
 * Default find function, retreive a config from given name
 *
 * @param {String} name wanted config
 * @param {String} complete true if we need to complete existing config with new
 * @return {Boolean} true if all is ok falser otherwise
 */
Config.prototype.find = function (name, complete) {
  // Message
  this.logger.debug([ '[ Config.find ] - Try to enable config for', name ].join(' '));

  // Search item
  var item = _.has(this.schemaList, name);

  // Has item ?
  if (item) {
    // Get value
    item = this.schemaList[name];

    // Create default object
    var obj = _.set({}, name, item);

    // Express item ??
    if (name === 'express' && _.has(obj, 'express')) {
      // Change depth assign
      obj = obj.express;
    }

    // Mongoose or sequelize item ?? transform to db
    if (name === 'mongoose' || name === 'sequelize') {
      // Change object name for db
      obj = {
        db : obj[name]
      };
    }

    // Add item at the end of list ?
    if (_.isBoolean(complete) && complete) {
      // Not extend but merge current object
      _.merge(this.schema, obj);
    } else {
      // Simple assignation
      this.schema = obj;
    }

    // Message
    this.logger.info([ '[ Config.find ] -', name, 'config was correcty activated.' ].join(' '));

    // Valid statement
    return true;
  }

  // Default statement
  return false;
};

/**
 * Enable Express config
 *
 * @param {Boolean} complete true if we need to append new config after existing
 * @return {Boolean} true if all is ok falser otherwise
 */
Config.prototype.enableExpress = function (complete) {
  // Process
  return this.enableSchema('express', complete);
};

/**
 * Enable Yocto Render config
 *
 * @param {Boolean} complete true if we need to append new config after existing
 * @return {Boolean} true if all is ok falser otherwise
 */
Config.prototype.enableRender = function (complete) {
  // Process
  return this.enableSchema('render', complete);
};

/**
 * Enable Yocto Router config
 *
 * @param {Boolean} complete true if we need to append new config after existing
 * @return {Boolean} true if all is ok falser otherwise
 */
Config.prototype.enableRouter = function (complete) {
  // Process
  return this.enableSchema('router', complete);
};

/**
 * Enable Mongoose config
 *
 * @param {Boolean} complete true if we need to append new config after existing
 * @return {Boolean} true if all is ok falser otherwise
 */
Config.prototype.enableMongoose = function (complete) {
  // Process
  return this.enableSchema('mongoose', complete);
};

/**
 * Enable PassportJs config
 *
 * @param {Boolean} complete true if we need to append new config after existing
 * @return {Boolean} true if all is ok falser otherwise
 */
Config.prototype.enablePassportJs = function (complete) {
  // Process
  return this.enableSchema('passportJs', complete);
};

/**
 * Enable Specific schema config
 *
 * @param {String} name default name to find in schema
 * @param {Boolean} complete true if we need to add new config after existing
 * @return {Boolean} true if all is ok falser otherwise
 */
Config.prototype.enableSchema = function (name, complete) {
  // Force complete to be a boolean
  complete = _.isBoolean(complete) ? complete : false;

  // Process
  return this.find(name, complete);
};

/**
 * Add a custom schema on config
 *
 * @param {String} name config name to add in schema
 * @param {Object} value config value to add in schema
 * @param {Boolean} enable true if we need to add auto enable of config
 * @param {Boolean} complete true if we need to add new config after existing
 * @return {Boolean} indicate success of this call
 */
Config.prototype.addCustomSchema = function (name, value, enable, complete) {
  // Schema already exists ?
  if (_.has(this.schemaList, name) || !_.isObject(value)) {
    // Error schema already exists
    this.logger.error([ '[ Config.addCustomerSchema ] - Cannot add new custom schema. Schema :',
      name, 'already exist, or given value is invalid.' ].join(' '));

    // Invalid schema
    return false;
  }

  // Add new schema
  _.extend(this.schemaList, _.set({}, name, value));

  // Auto enable ?
  if (_.isBoolean(enable) && enable) {
    // Enable schema
    return this.enableSchema(name, complete);
  }

  // Valid statement
  return _.has(this.schemaList, name);
};

/**
 * Default set function, a value to a specific params
 *
 * @param {String} name current name to use
 * @param {String} value current value to assign on params name
 * @return {Boolean} true if all is ok false otherwise
 */
Config.prototype.set = function (name, value) {
  // Check requirements
  if (!_.isUndefined(name) && _.isString(name) && !_.isEmpty(name)) {
    // Need to normalize path ?
    if (name === 'base') {
      // Is relative path ?
      if (!path.isAbsolute(value)) {
        // Normalize
        value = path.normalize([ process.cwd(), value ].join('/'));
      }
    }

    // Assign value
    this[name] = value;

    // Valid statement
    return true;
  }

  // Warn message
  this.logger.warning([ '[ Core.set ] - Invalid value given.',
    'name must be a string and not empty. Operation aborted !' ].join(' '));


  // Invalid statement
  return false;
};

/**
 * Retreive default configuration
 *
 * @return {Object} loaded object
 */
Config.prototype.getConfig = function () {
  // Default statement
  return this.get('config');
};

/**
 * Set config path
 *
 * @param {String} path default path to use
 * @return {Boolean} true if all is ok false otherwise
 */
Config.prototype.setConfigPath = function (path) {
  // Default statement
  return this.set('base', path);
};

/**
 * Return correct property from given name
 *
 * @param {String} name the property name
 * @return {Mixed} needed data
 */
Config.prototype.get = function (name) {
  // Default instance
  return this[name];
};

/**
 * Reload config from path
 *
 * @param {String} base if base exists and is valid reassign base and reload
 * @return {Boolean} true if load succeed false otherwise
 */
Config.prototype.reload = function (base) {
  // Check base before load for conditional assignation
  if (_.isString(base) && !_.isEmpty(base)) {
    // Change base
    this.base = base;
  }

  // Return load statement
  return this.load();
};

/**
 * Load password schema for current configuration
 *
 * @return {Object} return current promise
 */
Config.prototype.loadPassport = function () {
  // Retreive passport schema
  // this.schema = this.passport.get();
  // default statement
  return this.load();
};

/**
 * Auto enable validators schema for given list
 *
 * @param {Array} items array of items to enable
 * @return {Boolean} true if all is ok false otherwise
 */
Config.prototype.autoEnableValidators = function (items) {
  // Is valid format ?
  if (!_.isArray(items) || _.isEmpty(items)) {
    // Warn message invalid data
    this.logger.warning([ '[ Config.autoEnableValidator ] - Cannot check items.',
      'Is not an array or is empty.' ].join(' '));

    // Invalid statement
    return false;
  }

  // Parse item and check
  _.each(items, function (item) {
    // Schema exists ?
    if (_.has(this.schemaList, item)) {
      // Enable succeed ?
      if (!this.enableSchema(item, true)) {
        // Success or error ?
        this.logger.error([ '[ Config.autoEnableValidator ] - Auto enable [',
          item, '] failed' ].join(' '));
      }
    } else {
      // Warning
      this.logger.warning([ '[ Config.autoEnableValidator ] - Cannot enable [',
        item, '] schema does not exists' ].join(' '));
    }
  }.bind(this));

  // Default statement
  return true;
};

/**
 * Default load function, load data from all.js constant file
 *
 * @return {Object} return current promise
 */
Config.prototype.load = function () {
  // Create async process here
  var deferred = Q.defer();

  // Any errors ?? try/catch it
  try {
    // Default config object
    var config  = {};

    // Build pattern
    var pattern = path.normalize([ this.base, this.suffix, '*.json' ].join('/'));
    var penv    = path.normalize([ this.base, this.suffix, [
      this.env, '.json' ].join('')
    ].join('/'));

    // Get file (sync mode)
    var paths = glob.sync(pattern);

    // Has a valid path ? no ? stop process
    if (_.isEmpty(paths)) {
      throw 'No config files was found. Operation aborted !';
    }

    // Sort path
    paths = _.sortBy(paths, function (p) {
      var all         = path.normalize([ this.base, this.suffix, 'all.json' ].join('/'));
      var common      = path.normalize([ this.base, this.suffix, 'common.json' ].join('/'));
      var development = path.normalize([ this.base, this.suffix, 'development.json' ].join('/'));
      var staging     = path.normalize([ this.base, this.suffix, 'staging.json' ].join('/'));
      var production  = path.normalize([ this.base, this.suffix, 'production.json' ].join('/'));

      // Return data order
      return [ p === production, p === staging,
        p === development, p === common, p === all ].join('||');
    }.bind(this));

    // Parse all and merge if no error
    _.each(paths, function (path) {
      // Parse all file contains
      var item = JSON.parse(fs.readFileSync(path, 'utf-8'));

      // Merge data

      _.merge(config, item);

      // Has current env ? if test return false stop env was founded
      return penv !== path;
    });

    // Build schema
    this.schema = joi.object().required().keys(this.schema).unknown(true);

    // Validate !!!
    var result = joi.validate(config, this.schema, {
      abortEarly : false
    });

    // Has error
    if (!_.isNull(result.error)) {
      // Parse details
      _.each(result.error.details, function (error) {
        this.logger.warning([ '[ Config.load ] - Cannot update config an error occured. Error is :',
          utils.obj.inspect(error) ].join(' '));
      }.bind(this));

      // Throw exception error occured
      throw 'Config validation failed';
    }

    // Build directory value to UPPERCASE constants
    _.each(config.directory, function (dir) {
      // Build path
      var p = _.first(_.values(dir));

      // Is relative path ?
      if (!path.isAbsolute(p)) {
        p = path.normalize([ process.cwd(), p ].join('/'));
      }

      // Build key
      var k = [ _.first(_.keys(dir)), 'directory' ].join('_').toUpperCase();

      // Assign
      this[k] = p;
    }.bind(this));

    // Change state
    this.state = true;

    // Valid so assign config
    this.config = result.value;

    // Log message
    this.logger.info([ '[ Config.load ] - Success - Config file was changed with files based on :',
      [ this.base, this.suffix ].join('/') ].join(' '));

    // Resolve with valid value
    deferred.resolve(this.config);
  } catch (e) {
    // Error message
    this.logger.error([ '[ Config.load ] - an error occured during load config file. Error is :',
      e ].join(' '));

    // Reject
    deferred.reject(e);
  }

  // Return true if all is ok
  return deferred.promise;
};

// Default export
module.exports = function (l) {
  // Is a valid logger ?
  if (_.isUndefined(l) || _.isNull(l)) {
    logger.warning('[ Config.constructor ] - Invalid logger given. Use internal logger');

    // Assign
    l = logger;
  }

  // Default statement
  return new Config(l);
};
